ARM: nxp_imx: rt10xx: migrate ARM, AHB and IPG dividers to DT

mniestroj opened the pull request:
Those dividers were configured in Kconfig so far. Add 'arm-podf',
'ahb-podf' and 'ipg-podf' "fixed-factor-clock" compatible DT child nodes
under 'ccm' (Clock Control Module) and use configured 'clock-div' values
instead of Kconfig equivalents.

danieldegrasse commented on dts/bindings/clock/nxp,imx-ccm.yaml:
```diff
@@ -14,6 +14,21 @@ properties:
+    ipg-div:
+      type: int
+      required: true
+      description: IPG clock divider
```
When these values were encoded using Kconfigs, a range was enforced. Can you document the range here for future users?

mniestroj replied:
Sure, done.

bbilas commented on dts/arm/nxp/nxp_rt10xx.dtsi:
```diff
@@ -230,9 +230,13 @@
 
 		ccm: ccm@400fc000 {
-			compatible = "nxp,imx-ccm";
+			compatible = "nxp,imx-ccm", "nxp,imx-rt10xx-ccm";
```
`nxp,imx-ccm` is included within `nxp,imx-rt10xx-ccm.yaml` file so is it not enough to keep `nxp,imx-rt10xx-ccm` only?

<pre>
compatible = "nxp,imx-rt10xx-ccm";
</pre>

mniestroj replied:
`clock_control_mcux_ccm.c` driver relies on `nxp,imx-ccm`, which is shared across rt10xx and imx8mm families. Here we introduce `nxp,imx-rt10xx-ccm` bindings, which are "extension" just for rt10xx. Those addiitonal properties are used only by rt10xx SoC specific initialization code (at least for now), while all the other are generic enough for `clock_control_mcux_ccm.c` to work as it was before.

bbilas replied:
It's perfectly clear now, thanks.

galak reviewed (changes requested):
Can we represent these as `compatible = "fixed-factor-clock"`?

mniestroj committed:
ARM: nxp_imx: rt10xx: migrate ARM, AHB and IPG dividers to DT

Those dividers were configured in Kconfig so far. Add 'arm-podf',
'ahb-podf' and 'ipg-podf' "fixed-factor-clock" compatible DT child nodes
under 'ccm' (Clock Control Module) and use configured 'clock-div' values
instead of Kconfig equivalents.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj commented:
> Can we represent these as compatible = "fixed-factor-clock"?

@galak I reworked PR to use "fixed-factor-clock". Looks cleaner now, I think.

galak reviewed (dismissed):
LGTM

danieldegrasse commented on soc/arm/nxp_imx/rt/soc_rt10xx.c:
```diff
@@ -135,11 +135,14 @@ static ALWAYS_INLINE void clock_init(void)
 
-#ifdef CONFIG_HAS_ARM_DIV
-	CLOCK_SetDiv(kCLOCK_ArmDiv, CONFIG_ARM_DIV); /* Set ARM PODF */
+#if DT_NODE_EXISTS(DT_CHILD(CCM_NODE, arm_podf))
```
Could you use BUILD_ASSERT here for each clock to check the range information that was previously encoded via KConfig? I'd like to preserve this as it makes it easier for the user to avoid an invalid clock configuration, IMO.

danieldegrasse replied:
To add to this, I am also OK with use comments in the DTS to document this. Just want to make it easier for developers to find this information. Clock changes themselves LGTM, I verified this PR on all platforms except the RT1064 and RT1020

mniestroj replied:
@danieldegrasse I've followed your first suggestion: https://github.com/zephyrproject-rtos/zephyr/pull/49782#issuecomment-1290269083

mniestroj commented:
Rebased and added `BUILD_ASSERT_PODF_IN_RANGE` helper macro with build-time verification of `arm-podf`, `ahb-podf` and `ipg-podf`.

As an example if `ipg-podf` is set to 5 (which is out of 1-4 range), then following error occurs during build:
```
In file included from /zephyr/include/zephyr/toolchain.h:50,
                 from /zephyr/include/zephyr/kernel_includes.h:19,
                 from /zephyr/include/zephyr/kernel.h:17,
                 from /zephyr/soc/arm/nxp_imx/rt/soc_rt10xx.c:7:
/zephyr/soc/arm/nxp_imx/rt/soc_rt10xx.c: In function 'clock_init':
/zephyr/include/zephyr/toolchain/gcc.h:77:36: error: static assertion failed: "ipg_podf is out of supported range (1, 4)"
   77 | #define BUILD_ASSERT(EXPR, MSG...) _Static_assert(EXPR, "" MSG)
      |                                    ^~~~~~~~~~~~~~
/zephyr/soc/arm/nxp_imx/rt/soc_rt10xx.c:27:9: note: in expansion of macro 'BUILD_ASSERT'
   27 |         BUILD_ASSERT(DT_PROP(DT_CHILD(CCM_NODE, podf), clock_div) >= (a) && \
      |         ^~~~~~~~~~~~
/zephyr/soc/arm/nxp_imx/rt/soc_rt10xx.c:152:9: note: in expansion of macro 'BUILD_ASSERT_PODF_IN_RANGE'
  152 |         BUILD_ASSERT_PODF_IN_RANGE(ipg_podf, 1, 4);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
[81/160] Building C object zephyr/arch/arch/arm/core/aarch32/cortex_m/CMakeFiles/arch__arm__core__aarch32__cortex_m.dir/fault.c.obj
```
